cmd/serve_test.go

```diff
@@ -0,0 +1,75 @@
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"log"
+	"net"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// Thread safe buffer to avoid data races when setting a custom writer
+// for the log
+type Buffer struct {
+	b bytes.Buffer
+	m sync.Mutex
+}
+
+func (b *Buffer) Read(p []byte) (n int, err error) {
+	b.m.Lock()
+	defer b.m.Unlock()
+	return b.b.Read(p)
+}
+
+func (b *Buffer) Write(p []byte) (n int, err error) {
+	b.m.Lock()
+	defer b.m.Unlock()
+	return b.b.Write(p)
+}
+
+func (b *Buffer) String() string {
+	b.m.Lock()
+	defer b.m.Unlock()
+	return b.b.String()
+}
+
+func TestServe(t *testing.T) {
+	tempDir := t.TempDir()
+	buf := Buffer{}
+	log.SetOutput(&buf)
+	defer func() {
+		log.SetOutput(os.Stderr)
+	}()
+	ServeCmd.SetArgs([]string{"--data-dir", tempDir})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go ServeCmd.ExecuteContext(ctx)
+
+	if !waitForServer(":35354") {
+		assert.FailNow(t, "server did not start")
+	}
+	cancel()
```

Comment:
So this isn't the best idea, it'll leave the HTTP and SSH goroutines running in the background. `server.NewServer` doesn't take a context so I'll have to think of something different here, or maybe a new PR where `NewServer` takes a context that would stop all the server goroutines and exit cleanly when cancelled.

Reply:
Fixed in 3dbeec0